tss-esapi/examples/duplication.rs

```diff
@@ -0,0 +1,481 @@
+        .map_err(|err| {
+            eprintln!("⚠️  {}", err);
+            err
+        })
+        .unwrap();
+
+    // Do an hmac with it.
+    let hmac1 = context_1
+        .execute_with_nullauth_session(|ctx| {
+            let loaded_hmackey = ctx
+                .load(
+                    loaded_storage_key,
+                    hmac_key.out_private.clone(),
+                    hmac_key.out_public.clone(),
+                )
+                .unwrap();
+
+            ctx.execute_with_temporary_object(loaded_hmackey.into(), |ctx, handle| {
+                ctx.hmac(handle.into(), input_data.clone(), HashingAlgorithm::Sha256)
+            })
+        })
+        .unwrap();
+
+    // Great! Let's get to duplicating.
+
+    // We need the name of the object we are duplicating - for us, that's the storage_key.
+
+    let (_, object_to_duplicate_name, _) =
+        context_1.read_public(loaded_storage_key.into()).unwrap();
+
+    // Now, we can compute the real policy for the duplication.
+
+    let policy_auth_session = context_1
+        .execute_without_session(|ctx| {
+            let policy_auth_session = ctx
+                .start_auth_session(
+                    None,
+                    None,
+                    None,
+                    SessionType::Policy,
+                    SymmetricDefinition::AES_128_CFB,
+                    HashingAlgorithm::Sha256,
+                )
+                .expect("Start auth session failed")
+                .expect("Start auth session returned a NONE handle");
+
+            let (policy_auth_session_attributes, policy_auth_session_attributes_mask) =
+                SessionAttributesBuilder::new()
+                    .with_decrypt(true)
+                    .with_encrypt(true)
+                    .build();
+
+            ctx.tr_sess_set_attributes(
+                policy_auth_session,
+                policy_auth_session_attributes,
+                policy_auth_session_attributes_mask,
+            )
+            .expect("tr_sess_set_attributes call failed");
+
+            let policy_session = PolicySession::try_from(policy_auth_session)
+                .expect("Failed to convert auth session into policy session");
+
+            ctx.policy_auth_value(policy_session)
+                .expect("Policy auth value");
+
+            ctx.policy_command_code(policy_session, CommandCode::Duplicate)
+                .map(|_| policy_auth_session)
+
+            /*
+            ctx.policy_duplication_select(
+                policy_session,
+                object_to_duplicate_name,
+                target_parent_name,
+                false,
+            )
+            .map(|_| policy_auth_session)
+            */
+        })
+        .unwrap();
+
+    // ALWAYS FAILS!!!
+    // For some reason this has a policy error - I was copy pasting from the
+    // tests in duplication_commands_tests.rs
+    let public = storage_key.out_public.clone();
+    let (data, duplicate, secret) = context_1
```

Comment:
The only thing that differs is this separation. If both of these blocks are executed in one go, that is the PolicySession setup is followed by the `ctx.duplicate` call the test succeeds and that's how the stuff in other examples is laid out.

Is that sufficient for you to continue or were you asking me to debug it further? (that is: were you asking why just the `policy_auth_session` value is not sufficient for the policy to succeed).

Reply:
It seems I'd already refactored for that. I've updated the PR, but it still fails.

At this point I really actually feel quite lost and unable to solve this, so as much help as possible would be really appreciated :)

Reply:
The only difference that I have noticed when I looked at other examples is they do not have this "middle" key but uses the primary key directly as a storage key.

Reply:
But then what would "new parent name" be? A parent can't be duplicated iirc.

Reply:
I think I managed to figure out what's going wrong.

You don't need to set the command code, it's done automatically by a successful `PolicyDuplicationSelect`:

```
If the command succeeds, policySession→commandCode is set to TPM_CC_Duplicate.
```

But more importantly, you must only set the policy session on the context as the last step before executing the duplication command. At the moment you have an `load_external_public` in-between which then gets reflected in your policy digest. I'd recommend just running the duplication command under a `ctx.execute_with_session`.

I changed those two things and the example works for me now.

Reply:
Holy shit, that did it! THANK YOU!!!!!